File: app/src/main/java/com/github/weatherapp/injection/module/AppModule.java
package com.github.weatherapp.injection.module;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.github.weatherapp.core.AppSettings;
import com.github.weatherapp.data.SharedPrefsAppSettings;
import com.github.weatherapp.ui.resource.AndroidMessageResourceProvider;
import com.github.weatherapp.ui.resource.MessageResourceProvider;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

/*
    Android specific application module
 */
@Module
public class AppModule {
    private final Context appContext;

    public AppModule(Context appContext) {
        this.appContext = appContext.getApplicationContext();
    }

    @Provides
    @Singleton
    static MessageResourceProvider provideMessageResourceProvider(Context context) {
        return new AndroidMessageResourceProvider(context);
    }

    @Provides
    @Singleton
    static AppSettings provideAppSettings(SharedPreferences preferences) {
        return new SharedPrefsAppSettings(preferences);
    }

    @Provides
    @Singleton
    static SharedPreferences provideSharedPreferences(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context);
    }

    @Provides
    @Singleton
    Context provideAppContext() {
        return appContext;
    }
}

File: app-kotlin/src/main/kotlin/com/github/weatherapp/injection/module/AppModule.kt
package com.github.weatherapp.injection.module

import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager
import com.github.weatherapp.core.AppSettings
import com.github.weatherapp.data.SharedPrefsAppSettings
import com.github.weatherapp.ui.resource.AndroidMessageResourceProvider
import com.github.weatherapp.ui.resource.MessageResourceProvider
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

/*
    Android specific application module
 */
@Module
class AppModule(appContext: Context) {
    private val appContext: Context

    init {
        this.appContext = appContext.applicationContext
    }

    @Provides
    @Singleton
    fun provideAppContext(): Context {
        return appContext
    }

        @Provides
        @Singleton
        fun provideMessageResourceProvider(context: Context): MessageResourceProvider {
            return AndroidMessageResourceProvider(context)
        }

        @Provides
        @Singleton
        fun provideAppSettings(preferences: SharedPreferences): AppSettings {
            return SharedPrefsAppSettings(preferences)
        }

        @Provides
        @Singleton
        fun provideSharedPreferences(context: Context): SharedPreferences {
            return PreferenceManager.getDefaultSharedPreferences(context)
        }
}
